### PR TITLE
Add missing source spans for messages

### LINKED ISSUE #59: Add Source Spans to Remaining Messages
## Description
Messages with source spans have been introduced in #45, but outside of the messages in the transformation functions, most messages have preliminary received `S.NoSrcSpan` as their source span, because I thought adding actual source spans would fit better in #46 or in its own issue / pull request. These missing source spans should now be added.
In addition, messages where no fitting source span is available should not cause HLint suggestions, so `S.NoSrcSpan` should be avoided in these cases. One way to this is to add an additional smart constructor for messages which does not take a source span.

### PR DESCRIPTION
**Pull request by stu204146-kiel**
### Issue
Closes #59.

### Description of the Change
I have added source spans to all messages where it was possible. When it was not possible to give an accurate source span, I have created a new smart constructor for messages that have no source spans.

### Verification Process

I have tested it manually with example programs.

### CODE REVIEW

File: src/lib/HST/CoreAlgorithm.hs [hunk 1]
```diff
@@ -85,50 +86,56 @@ match vars@(x : xs) eqs er
...
 
 -- | Creates a case expression that performs pattern matching on the variable
 --   bound by the given variable pattern.
 makeRhs :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
-        => S.Pat a   -- ^ The fresh variable pattern to match.
+        => S.SrcSpan a -- ^ The source span of the corresponding definition.
+        -> S.Pat a   -- ^ The fresh variable pattern to match.
         -> [S.Pat a] -- ^ The remaing fresh variable patterns.
         -> [Eqs a]   -- ^ The equations.
         -> S.Exp a   -- ^ The error expression for pattern-matching failures.
```
Comment (stu204767) [inline src/lib/HST/CoreAlgorithm.hs]:
```suggestion
        => S.SrcSpan a -- ^ The source span of the corresponding definition.
        -> S.Pat a     -- ^ The fresh variable pattern to match.
        -> [S.Pat a]   -- ^ The remaining fresh variable patterns.
        -> [Eqs a]     -- ^ The equations.
        -> S.Exp a     -- ^ The error expression for pattern-matching failures.
```

File: src/lib/HST/Util/Messages.hs [hunk 2]
```diff
@@ -35,6 +36,9 @@ data Message = Message { msgSeverity :: Severity
 message :: Severity -> S.SrcSpan a -> String -> Message
 message severity srcSpan = Message severity (S.toMsgSrcSpan srcSpan)
 
+messageWithoutSrcSpan :: Severity -> String -> Message
```
Comment (just95) [inline src/lib/HST/Util/Messages.hs:40]:
```suggestion
-- | Like 'message' but for messages without source span.
messageWithoutSrcSpan :: Severity -> String -> Message
```

File: src/lib/HST/CoreAlgorithm.hs [hunk 3]
```diff
@@ -138,14 +145,15 @@ makeRhs x xs eqs er = do
 --   every missing constructor.
 computeAlts
   :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
-  => S.Pat a   -- ^ The variable pattern that binds the matched variable.
-  -> [S.Pat a] -- ^ The remaing fresh variable patterns.
-  -> [Eqs a]   -- ^ The equations to generate alternatives for.
-  -> S.Exp a   -- ^ The error expression for pattern-matching failures.
+  => S.SrcSpan a -- ^ The source span of the corresponding definition.
+  -> S.Pat a     -- ^ The variable pattern that binds the matched variable.
+  -> [S.Pat a]   -- ^ The remaing fresh variable patterns.
```
Comment (stu204767) [inline src/lib/HST/CoreAlgorithm.hs]:
```suggestion
  -> [S.Pat a]   -- ^ The remaining fresh variable patterns.
```

File: src/lib/HST/CoreAlgorithm.hs [hunk 4]
```diff
@@ -284,12 +294,13 @@ compareCons = (==) `on` getMaybePatConName
 --   once sharing is implemented.
 computeAlt
   :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
-  => S.Pat a   -- ^ The variable pattern that binds the matched variable.
+  => S.SrcSpan a -- ^ The source span of the corresponding definition.
+  -> S.Pat a   -- ^ The variable pattern that binds the matched variable.
   -> [S.Pat a] -- ^ The remaing fresh variable patterns.
   -> S.Exp a   -- ^ The error expression for pattern-matching failures.
   -> [Eqs a]   -- ^ A group of equations (see 'groupByCons').
```
Comment (stu204767) [inline src/lib/HST/CoreAlgorithm.hs]:
```suggestion
  => S.SrcSpan a -- ^ The source span of the corresponding definition.
  -> S.Pat a     -- ^ The variable pattern that binds the matched variable.
  -> [S.Pat a]   -- ^ The remaining fresh variable patterns.
  -> S.Exp a     -- ^ The error expression for pattern-matching failures.
  -> [Eqs a]     -- ^ A group of equations (see 'groupByCons').
```

File: src/lib/HST/Util/Messages.hs [hunk 5]
```diff
@@ -35,6 +36,9 @@ data Message = Message { msgSeverity :: Severity
...
+messageWithoutSrcSpan severity = Message severity Nothing
```
Comment (stu204767) [inline src/lib/HST/Util/Messages.hs:41]: I would have chosen a shorter name (`messageNoSrc` for example), but this is fine as well. You should add a Haddock comment for this function, though.
Edit: @just95 fixed this two minutes before I submitted my review, what a coincidence!

**Review by just95 — APPROVED** [review]

**Review by stu204767 — CHANGES_REQUESTED** [review]
Very nice changes!
I will approve this when you add the missing Haddock comment.

**Review by stu204767 — APPROVED** [review]

**Pull request merged**